src/lib/util/graceful-shutdown.ts

```diff
@@ -1,7 +1,60 @@
-import { Logger } from '../logger';
+import { Logger, LogProvider } from '../logger';
 import { IUnleash } from '../types/core';
+import { GenericUnleashError } from '../error/unleash-error';
 
-function registerGracefulShutdown(unleash: IUnleash, logger: Logger): void {
+export interface IGracefulShutdown {
+    registerGracefulShutdownHook(
+        serviceName: string,
+        hook: () => Promise<void>,
```

Comment:
Maybe this helps in case we want to register a non-promise shutdown hook. I noticed a lot of `async` shutdown hooks that did not need to be `async` since they did not `await` anything.
```suggestion
        hook: () => void | Promise<void>,
```

Reply:
Agreed, added in 939054192. As well as a test proving that we can add and execute synchronous hooks as well